source/common/grpc/subscription_impl.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include "envoy/config/subscription.h"
+#include "envoy/event/dispatcher.h"
+
+#include "common/grpc/async_client_impl.h"
+
+#include "api/base.pb.h"
+
+namespace Envoy {
+namespace Grpc {
+
+template <class ResponseType, class ResourceType>
+class SubscriptionImpl : public Config::Subscription<ResponseType, ResourceType>,
+                         AsyncClientCallbacks<ResponseType> {
+public:
+  SubscriptionImpl(const envoy::api::v2::Node& node, Upstream::ClusterManager& cm,
+                   const std::string& remote_cluster_name, Event::Dispatcher& dispatcher,
+                   const google::protobuf::MethodDescriptor& service_method)
+      : SubscriptionImpl(node, new AsyncClientImpl<envoy::api::v2::DiscoveryRequest, ResponseType>(
+                                   cm, remote_cluster_name),
+                         dispatcher, service_method) {}
+
+  SubscriptionImpl(const envoy::api::v2::Node& node,
+                   AsyncClient<envoy::api::v2::DiscoveryRequest, ResponseType>* async_client,
+                   Event::Dispatcher& dispatcher,
+                   const google::protobuf::MethodDescriptor& service_method)
+      : async_client_(async_client), service_method_(service_method),
+        retry_timer_(dispatcher.createTimer([this]() -> void { establishNewStream(); })) {
+    request_.mutable_node()->CopyFrom(node);
+  }
+
+  ~SubscriptionImpl() {
+    if (stream_ != nullptr) {
+      stream_->reset();
+    }
+  }
+
+  void setRetryTimer() { retry_timer_->enableTimer(std::chrono::milliseconds(RETRY_DELAY_MS)); }
+
+  void establishNewStream() {
+    stream_ = async_client_->start(service_method_, *this, Optional<std::chrono::milliseconds>());
+    if (stream_ == nullptr) {
+      // TODO(htuch): Track stats and log failures.
+      setRetryTimer();
+      return;
+    }
+    sendDiscoveryRequest();
+  }
+
+  void sendDiscoveryRequest() {
+    if (stream_ == nullptr) {
+      return;
```

Comment:
Yeah. This occurs if a gRPC level issue arises and we are in the retry period to establish a new stream, the caller does an `updateResources()` and we end up here. This is covered by the `StreamCreationFailure` test.